selfservice/strategy/code/strategy.go

```diff
@@ -82,14 +119,348 @@ func NewStrategy(deps strategyDependencies) *Strategy {
+		for _, n := range container.NewFromJSON("", node.CodeGroup, data, "traits").Nodes {
+			// we only set the value and not the whole field because we want to keep types from the initial form generation
+			f.GetUI().GetNodes().SetValueAttribute(n.ID(), n.Attributes.GetValue())
+		}
+	} else if f.GetFlowName() == flow.LoginFlow {
+		// on Login flow we need to populate the form with the values from the initial form generation
+		for _, n := range container.NewFromJSON("", node.DefaultGroup, data, "").Nodes {
+			f.GetUI().GetNodes().SetValueAttribute(n.ID(), n.Attributes.GetValue())
+		}
+	}
+
+	return nil
+}
+
+type (
+	FlowType interface {
+		*login.Flow | *registration.Flow | *recovery.Flow | *verification.Flow
+		flow.Flow
+	}
+	FlowPayload interface {
+		*updateLoginFlowWithCodeMethod | *updateRegistrationFlowWithCodeMethod | *updateRecoveryFlowWithCodeMethod | *updateVerificationFlowWithCodeMethod
+	}
+	CreateCodeState[T FlowType, P FlowPayload]           func(context.Context, T, *Strategy, P) error
+	ValidateCodeState[T FlowType, P FlowPayload]         func(context.Context, T, *Strategy, P) error
+	AlreadyValidatedCodeState[T FlowType, P FlowPayload] func(context.Context, T, *Strategy, P) error
+	CodeStateManager[T FlowType, P FlowPayload]          struct {
+		f                         T
+		payload                   P
+		strategy                  *Strategy
+		createCodeState           CreateCodeState[T, P]
+		verifyCodeState           ValidateCodeState[T, P]
+		alreadyValidatedCodeState AlreadyValidatedCodeState[T, P]
+	}
+)
+
+func NewCodeStateManager[T FlowType, P FlowPayload](f T, s *Strategy, payload P) *CodeStateManager[T, P] {
+	return &CodeStateManager[T, P]{
+		f:        f,
+		strategy: s,
+		payload:  payload,
+	}
+}
+
+func (c *CodeStateManager[T, P]) SetCreateCodeHandler(fn CreateCodeState[T, P]) {
+	c.createCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) SetCodeVerifyHandler(fn ValidateCodeState[T, P]) {
+	c.verifyCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) SetCodeDoneHandler(fn AlreadyValidatedCodeState[T, P]) {
+	c.alreadyValidatedCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) validatePayload(ctx context.Context) error {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		if len(v.Identifier) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/identifier", "identifier"))
+		}
+	case *updateRegistrationFlowWithCodeMethod:
+		if len(v.Traits) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/traits", "traits"))
+		}
+	case *updateRecoveryFlowWithCodeMethod:
+		if len(v.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	case *updateVerificationFlowWithCodeMethod:
+		if len(v.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	default:
+		return errors.WithStack(herodot.ErrBadRequest.WithReason("received unexpected flow payload type"))
+	}
+	return nil
+}
```

Comment:
If there are no overlaps in the payload validation, there's not really a reason to move this into its own function with type assertion. It's easier to read and understand if it's "simply" in the code at the right place.